### DISCUSSION TITLE
union type prevent entity as selected option in Twig/FormExtension

### DISCUSSION THREAD

**Discussion by edejourmonka** [symfony/symfony#45933] [Q&A]
Since 6.0 the function is typed with string|array
`function twig_is_selected_choice(ChoiceView $choice, string|array$selectedValue): bool`

This prevent to have object in choices when used with EntityTypeField. Because object is needed by :
‘Symfony\Bridge\Doctrine\Form\ChoiceList\IdReader::getIdValue()’ requires an object. 

So I think object should be an accepted type for $selectedValue.
`function twig_is_selected_choice(ChoiceView $choice, string|array|object $selectedValue): bool`

At least, when I apply this change, it works fine.

For information, I load the choices on EntityTypeField because I use Select2 with an API call as it’s on a +40k rows entity. So on edit I want to have only one option : the current one.

**Comment by anDeRanV2** [comment]
If anyone else is facing the same issue, the solution is very simple - you need to add the `__toString` magic method to the class you are providing for EntityType.

Example:
```php
namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;
use Stringable;

class YourClassName implements Stringable
{

	// Your properties & methods

	public function __toString(): string
	{
	    return (string)$this->id;
	}
}
```

**Comment by stof** [comment]
To me, this looks like a mistake in the type added in `twig_is_selected_choice`

**Comment by stof** [comment]
hmm, actually no. This Twig test is used with the _value_ of the choice. and values are always strings.

Please provide a reproducing case here. I suspect that this might be related to custom usages of the function that don't use it in the expected way.
